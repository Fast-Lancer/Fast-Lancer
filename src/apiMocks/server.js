import { rest } from 'msw'
import { setupServer } from 'msw/node'

const url = process.env.REACT_APP_SUPABASE_URL + '/rest/v1'

let postedClient = {}
let postedProject = {}

export const server = setupServer(
  rest.get(url + '/clients', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(select === '*' && !id) {
      // response for getClients()
      return res(
        ctx.json([
          {
            id: 1,
            client_name: 'bob1',
            business_name: 'business1'
          },
          {
            id: 2,
            client_name: 'bob2',
            business_name: 'business2'
          },
          {
            id: 3,
            client_name: 'bob3',
            business_name: 'business3'
          }
        ])
      )
    } else if(id) {
      // response for getClient(id)
      if(id === 'eq.42') {
        // Janky workaround to make testing edit/create possible.
        // The post route mock will set the id to 42
        // So this responds with the posted data when client.id is 42
        return res(
          ctx.json(
            postedClient
          )
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              client_name: 'bob1',
              business_name: 'business1'
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/clients', (req, res, ctx) => {
    let [client] = req.body
    client = { ...client, id: 42 }
    postedClient = client
    return res(
      ctx.json([
        client
      ])
    )
  }),
  rest.get(url + '/projects', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(!id) {
      // response for getProjects()
      return res(
        ctx.json([
          {
            id: 1,
            title: 'project1',
            client_id: '1',
            clients: {
              client_name: 'bob1'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          },
          {
            id: 2,
            title: 'project2',
            client_id: '2',
            clients: {
              client_name: 'bob2'
            },
            
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          
          },
          {
            id: 3,
            title: 'project3',
            client_id: '3',
            clients: {
              client_name: 'bob3'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          }
        ])
      )
    } else if(id) {
      // response for getProjectById(id)
      if(id === 'eq.42') {
        return res(
          ctx.json(postedProject)
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              title: 'project1',
              client_id: '1',
              clients: {
                client_name: 'bob1'
              },
              date_start:'2022-01-20',
              date_end:'2022-01-21',
              URL: 'something@something.com',
              description: 'ok',
              notes: 'test notes',
              hourly_rate: 33,
              hours_quoted: 22,
              price_quoted: 300
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/projects', (req, res, ctx) => {
    let project = req.body
    project = { ...project, id: 42 }
    postedProject = project
    return res(
      ctx.json([project])
    )
  })
)
