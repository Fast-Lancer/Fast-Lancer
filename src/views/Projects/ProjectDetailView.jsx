import { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { useParams } from 'react-router-dom'
import Button from '../../components/Button/Button.jsx'
import ProjectDetail from '../../components/ProjectDetail/ProjectDetail.jsx'
import Title from '../../components/Title/Title.jsx'
import { getProjectById } from '../../services/projects.js'

export default function ProjectDetailView() {
  const { id } = useParams()
  const [project, setProject] = useState({})
  const [loading, setLoading] = useState(true)
  
  useEffect(() => {
    getProjectById(id).then(data => setProject(data)).finally(() => setLoading(false))
  }, [])

  if(loading) return <h1>Loading...</h1>
  return (
    <div>
      <Title pageTitle='project detail' pageHeader={project.title} />
      <ProjectDetail project={project} />
    </div>
  )
}
