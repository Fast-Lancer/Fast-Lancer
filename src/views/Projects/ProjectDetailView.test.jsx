import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import ProjectDetailView from './ProjectDetailView.jsx'
import { server } from '../../apiMocks/server.js'

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})


it('should display a loading screen, then a project detail view', async () => {

  const { container } = render(
    <MemoryRouter initialEntries={['/projects/1']}>
      <ProjectDetailView />
    </MemoryRouter>
  )

  
  expect(screen.getByText(/loading/gi)).toBeInTheDocument()
  
  await screen.findByText(/bob1/gi)
  await screen.findByText(/33/gi)
  await screen.findByText(/something@something.com/gi)
  await screen.findByText(/project1/gi)
  await screen.findByText(/2022-01-20/gi)
  await screen.findByText(/2022-01-21/gi)
  await screen.findByText(/ok/gi)
  await screen.findByText(/test notes/gi)

  expect(container).toMatchSnapshot()
})
