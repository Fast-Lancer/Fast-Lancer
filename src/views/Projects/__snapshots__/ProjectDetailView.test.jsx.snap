// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a loading screen, then a project detail view 1`] = `
<div>
  <div>
    <div
      class="titleBody"
    >
      <section
        class="title"
      >
        project1
      </section>
      <ul
        class="tabnav"
      >
        <li
          class="active"
        >
          <a
            href="/projects"
          >
            Projects
          </a>
        </li>
        <li>
          <a
            href="/clients"
          >
            Clients
          </a>
        </li>
        <li>
          <a
            href="/projects/edit/undefined"
          >
            Edit Project
          </a>
        </li>
      </ul>
    </div>
    <div
      class="projectDetail"
    >
      <div
        class="clientName"
      >
        <h1
          aria-label="client-name"
        >
          bob1
        </h1>
      </div>
      <section
        class="infoSection"
      >
        <div
          aria-label="hourly-rate"
        >
          Hourly Rate: 
          33
        </div>
        <div
          aria-label="URL"
        >
          Repo: 
          something@something.com
        </div>
        <div
          aria-label="start-date"
        >
          Start Date: 
          2022-01-20
        </div>
        <div
          aria-label="end-date"
        >
          End Date: 
          2022-01-21
        </div>
      </section>
      <section
        class="alignLeft"
      >
        <h2>
          Description
        </h2>
        <p
          aria-label="description"
        >
          ok
        </p>
      </section>
      <section
        class="alignLeft"
      >
        <h2>
          Notes
        </h2>
        <p
          aria-label="notes"
        >
          test notes
        </p>
      </section>
    </div>
  </div>
</div>
`;
